### PR TITLE
pagination is added

### PR DESCRIPTION
**Pull request by DRITE**
extracted from #11

### CODE REVIEW

File: ui/src/components/ui/Pagination/Pagination.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import styles from './Pagination.module.css'
+import { $do } from '../../../actions/types'
+import { WidgetTableMeta } from '@tesler-ui/core/interfaces/widget'
+import { AppState } from '../../../interfaces/storeSlices'
+
+export interface PaginationProps {
+    meta: WidgetTableMeta
+}
+
+const Pagination: React.FC<PaginationProps> = props => {
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/components/ui/Pagination/Pagination.tsx:13]: Haven't noticed that, thanks

File: ui/src/components/ui/Pagination/Pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    const handlePageChange = React.useCallback(
+        (page: number) => {
+            dispatch($do.bcChangePage({ bcName, page }))
+        },
+        [dispatch, bcName]
+    )
+
+    return (
+        <>
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/components/ui/Pagination/Pagination.tsx:29]: ok

File: ui/src/components/ui/Pagination/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            {!!total && (
+                <AntPagination
+                    className={styles.pagination}
+                    size="small"
+                    pageSize={limit}
+                    defaultCurrent={page}
+                    current={page}
+                    total={total}
+                    onChange={page => {
+                        handlePageChange(page)
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/components/ui/Pagination/Pagination.tsx:39]: exactly

File: ui/src/epics/view.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { $do } from '../actions/types'
+import { BcCountResponse } from '../interfaces/bcCount'
+import { buildBcUrl } from '@tesler-ui/core'
+import { fetchBcCount } from '../api/bcCount'
+
+const bcFetchCountEpic: CustomEpic = (action$, store) =>
+    action$
+        .ofType(actionTypes.bcFetchDataRequest)
+        .mergeMap(action => {
+            const bcName = action.payload.bcName as string
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/epics/view.ts:12]: I guess it's not a good idea to use deprecated field in demo. I'll just get bcName from widget.

File: ui/src/epics/view.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            const state = store.getState()
+            const filters = state.screen.filters[bcName]
+            const bcUrl = buildBcUrl(bcName)
+            return fetchBcCount(bcUrl, filters).mergeMap(({ data }: BcCountResponse) =>
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/epics/view.ts:16]: ok

File: ui/src/reducers/view.ts [hunk 6]
```diff
@@ -1,11 +1,15 @@
-import { AnyAction } from '../interfaces/actions'
+import { actionTypes, AnyAction } from '../interfaces/actions'
 import { AppState } from '../interfaces/storeSlices'
 import { ViewState } from '@tesler-ui/core/interfaces/view'
 
+export type CustomView = ViewState & {
+    bcRecordsCount: Record<string, { count: number }>
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/reducers/view.ts:6]: exactly

File: ui/src/components/ui/Pagination/Pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    meta: WidgetTableMeta
+}
+
+const Pagination: React.FC<PaginationProps> = props => {
+    const dispatch = useDispatch()
+
+    const { bcName, limit: metaLimit } = props.meta
+    const { limit: bcLimit, page } = useSelector((state: AppState) => state.screen.bo.bc[bcName])
+    const total = useSelector((state: AppState) => state.view.bcRecordsCount[bcName]?.count)
+    const limit = metaLimit || bcLimit
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/components/ui/Pagination/Pagination.tsx:19]: ok

[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/components/ui/Pagination/Pagination.tsx:19]: [By default](https://github.com/tesler-platform/tesler/blob/master/tesler-core/src/main/java/io/tesler/core/dto/data/view/WidgetDTO.java#L139) `metaLimit ` is 0
 In my opinion it's a wrong way to use zero value  from `metaLimit`. In this case we should set `bcLimit` to 0.

[the remark this answers is not in the document]
Comment (Dergash) [reply ui/src/components/ui/Pagination/Pagination.tsx:19]: What if some specific widget needs 0 limit but another widget needs 5?
E.g.:
1) First view, some widget with 0 metaLimit. Idk why, may be to fetch row meta and operations without fetching the data
2) Second view, another widget of the same business component who needs data.
Not possible with just `bcLimit`

File: ui/src/api/bcCount.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { buildUrl, axiosGet } from '@tesler-ui/core'
+import qs from 'query-string'
+import { BcCountParamsMap, BcCountResponse } from '../interfaces/bcCount'
+
+export function fetchBcCount(bcName: string, params: BcCountParamsMap = {}) {
+    const url = buildUrl`count/dashboard/` + bcName
+    const stringParams = qs.stringify(params, { encode: true })
```
[the remark this answers is not in the document]
Comment (DRITE) [reply ui/src/api/bcCount.ts]: I've got it

File: ui/src/components/ui/Pagination/Pagination.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dergash) [reply ui/src/components/ui/Pagination/Pagination.tsx:19]: We can. Whether we set `limit` in widget.json to null or zero, DTO maps this field to `0` which is what we receive on our side.

**Review by Dergash — APPROVED** [review]

**Pull request merged**